### PR TITLE
1552136: Convert CounterStorageEngineTests.kt to Rust

### PR DESCRIPTION
**Pull request by mdboom**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/tests/counter.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            ..Default::default()
+        });
+
+        metric.add(&glean, 1);
+
+        let snapshot = StorageManager
+            .snapshot_as_json(glean.storage(), "store1".into(), true)
+            .unwrap();
+        assert_eq!(
+            "{\"counter\":{\"telemetry.counter_metric\":1}}",
```
Comment (badboy) [inline glean-core/tests/counter.rs]: Instead of constructing a string we should be able to build the JSON value like:

```
json!({"counter": {"telemetry.counter_metric": 1}})
```

and compare it to  `snapshot`.

File: glean-core/tests/counter.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
+
+        let metric = CounterMetric::new(CommonMetricData {
+            name: "counter_metric".into(),
+            category: "telemetry".into(),
+            send_in_pings: vec!["store1".into()],
+            disabled: false,
+            lifetime: Lifetime::User,
+            ..Default::default()
```
Comment (badboy) [inline glean-core/tests/counter.rs]: The default here is unneeded, we're filling all fields now.

File: glean-core/tests/common/mod.rs [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+pub fn tempdir() -> (tempfile::TempDir, String) {
+    let t = tempfile::tempdir().unwrap();
+    let name = t.path().display().to_string();
+    (t, name)
+}
+
+pub const GLOBAL_APPLICATION_ID: &str = "org.mozilla.glean.test.app";
```
Comment (badboy) [inline glean-core/tests/common/mod.rs]: Creating the `glean` object is the same across all tests.
In `tests/metrics.rs` we have;

```rust
fn new_glean() -> (Glean, TempDir) {
    let dir = tempfile::tempdir().unwrap();
    let tmpname = dir.path().display().to_string();

    let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();

    (glean, dir)
}
```

Can we use that everywhere instead of just the `tempdir` function?

File: glean-core/tests/counter.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        assert_eq!(
+            "{\"counter\":{\"telemetry.counter_metric\":1}}",
+            ::serde_json::to_string(&snapshot).unwrap(),
+        );
+    }
+
+    // Make a new glean instance here, which should force reloading of the data from disk
+    // so we can ensure it persisted, because it has User lifetime
+    {
+        let glean = Glean::new(&tmpname, GLOBAL_APPLICATION_ID).unwrap();
```
Comment (badboy) [inline glean-core/tests/counter.rs]: Ah, we need a new glean object (re `new_glean` from above).
Maybe we expose both things and in this test create the Glean objects ourselves, but otherwise use `new_glean` taking care of that?

File: glean-core/tests/counter.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    metric.add(&glean, 1);
+
+    for store_name in store_names {
+        let snapshot = StorageManager
+            .snapshot_as_json(glean.storage(), &store_name, true)
+            .unwrap();
+
+        assert_eq!(1, snapshot.as_object().unwrap().len());
+        println!("{:?}", snapshot);
```
Comment (badboy) [inline glean-core/tests/counter.rs]: Stray `println` in test code.

File: glean-core/tests/counter.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        assert_eq!(1, snapshot["counter"]["telemetry.counter_metric"]);
```
Comment (badboy) [inline glean-core/tests/counter.rs]: oooooh, that ... works.

We should use the same style everywhere though. I think building the JSON object as above (`json!(...)`) would be sensible.

File: glean-core/tests/counter.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    let snapshot = StorageManager.snapshot(glean.storage(), "store1", true);
+    assert!(snapshot.is_some());
+    // Check that getting a new snapshot for "store1" returns an empty store.
+    assert!(StorageManager
+        .snapshot(glean.storage(), "store1", false)
+        .is_none());
+    // Check that we get the right data from both the stores. Clearing "store1" must
+    // not clear "store2" as well.
+    let snapshot2 = StorageManager.snapshot(glean.storage(), "store2", true);
+    assert!(snapshot.unwrap().len() != 0);
```
Comment (badboy) [inline glean-core/tests/counter.rs]: Instead of checking the length, we can just check that it `is_some`, just as above.
Also checking `snapshot` again seems weird, that check was done in line 111.

File: glean-core/tests/counter.rs (hunk 1 above)
Comment (mdboom) [reply glean-core/tests/counter.rs]: In this case I actually think it's valuable to test against a string -- we want to make sure we're serializing with no unnecessary whitespace and in a single line etc.  Could move that out into its own test instead, of course...

File: glean-core/tests/counter.rs (hunk 2 above)
Comment (mdboom) [reply glean-core/tests/counter.rs]: Make sense.

File: glean-core/tests/common/mod.rs (hunk 3 above)
Comment (mdboom) [reply glean-core/tests/common/mod.rs]: Sure, that's a good idea.  Do you think that belongs in this file or in `common/mod.rs`?

Comment (badboy) [reply glean-core/tests/common/mod.rs]: If we re-use it in mod.rs.

File: glean-core/tests/counter.rs (hunk 1 above)
Comment (badboy) [reply glean-core/tests/counter.rs]: But the string-serialization happens in this test, not in library code. We therefore test only that `serde_json::to_string` kinda does what we expect, but don't test that anything internally actually uses that.

File: glean-core/tests/counter.rs (hunk 7 above)
Comment (mdboom) [reply glean-core/tests/counter.rs]: It's not quite checking the same thing, though.  The function could easily return `Some(Object::new())` which would pass `is_some()`, but fail `unwrap().len() != 0`.  I think it's important to assert that *if* it returns `Some` that `Object` also has at least one entry.

Reordering things makes sense, though.

File: glean-core/tests/counter.rs (hunk 1 above)
Comment (mdboom) [reply glean-core/tests/counter.rs]: Ah, good point. :slightly_frowning_face: Indeed, makes more sense to compare JSON objects then.

File: glean-core/tests/counter.rs (hunk 7 above)
Comment (badboy) [reply glean-core/tests/counter.rs]: Hm.... `snapshot` already has this check though and it serializes a HashMap. Even an empty hashmap is at least serialized as `{}`, therefore defeating your test.

IMO it would be much more sensible to test that we get `None` when we expect it and `Some` to indicate it contains data and to look into the data if needed.

(And at least `len() != 0` should be written as `!...is_empty()`)

Comment (mdboom) [reply glean-core/tests/counter.rs]: Got it -- I forgot that `snapshot` is snapshotting as a string.

File: Makefile [hunk 8]
```diff
@@ -12,7 +12,7 @@ emulator:
 .PHONY: install
 
 lint: fmt
-	cargo clippy --all
+	cargo clippy --all --all-targets --all-features -- -D warnings
```
Comment (badboy) [inline Makefile]: 🎉

File: glean-core/tests/common/mod.rs [hunk 9]
```diff
@@ -4,6 +4,10 @@
 
 use glean_core::Glean;
 
+// #[allow(dead_code)] is required on functions in this module as a workaround
+// for https://github.com/rust-lang/rust/issues/46379
+
+#[allow(dead_code)]
```
Comment (badboy) [inline glean-core/tests/common/mod.rs]: We should be able to put that globally for this file:

`#![allow(dead_code)]` as the first non-comment line in the file.

File: glean-core/tests/common/mod.rs [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use glean_core::Glean;
+
+// #[allow(dead_code)] is required on functions in this module as a workaround
+// for https://github.com/rust-lang/rust/issues/46379
```
Comment (mdboom) [inline glean-core/tests/common/mod.rs]: @badboy: The issue sort of makes sense given that test modules are compiled as separate crates. I still think the upside of putting this kind of code in a common place outweighs the downsides of having declare everything with #[allow(dead_code)], but if you have any alternative suggestions, let me know.

File: glean-core/tests/common/mod.rs [hunk 11]
```diff
@@ -2,12 +2,11 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+// #[allow(dead_code)] is required on this module as a workaround for
+// https://github.com/rust-lang/rust/issues/46379
+#[allow(dead_code)]
```
Comment (badboy) [inline glean-core/tests/common/mod.rs]: Missing the bang:

```
help: if you just forgot a `!`, use: `#![allow(dead_code)]`
```

File: glean-core/tests/common/mod.rs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// #[allow(dead_code)] is required on this module as a workaround for
+// https://github.com/rust-lang/rust/issues/46379
+#[allow(dead_code)]
```
Comment (badboy) [inline glean-core/tests/common/mod.rs]:
```suggestion
#![allow(dead_code)]
```
